rr_viz/src/rr_viz_tabs.py

```diff
@@ -80,46 +84,86 @@ def __init__(self, parent):
-
         # Get Battery levels 
         self.good_voltage = rospy.get_param("/battery/good_voltage", "31")
         self.ok_voltage = rospy.get_param("/battery/ok_voltage", "26")
         self.low_voltage = rospy.get_param("/battery/low_voltage", "22")
+        self.battery_state = False
+
+        # Set up topic subscribers
+        self.battery_level_sub_name = rospy.get_param("/battery/level", "/vesc_driver/battery")
+        self.battery_level_sub = rospy.Subscriber(self.battery_level_sub_name, BatteryState, self.battery_level_update)
 
-        # Variables required to detect if connected to ROS MASTER
-        self.ros_loss_triggered = True
-        self.rosMasterIP = '192.168.10.100'
-        self.subprocess_command = 'fping -nV -t 50 ' + self.rosMasterIP
-        self.fp_status = "alive"  
-
-        # Set up timer 
-        self.timer_period = 1500
-        self.timer = QTimer(self)
-        self.timer.timeout.connect(self.ros_is_up)
-        self.timer.start(self.timer_period)
-
-    def ros_is_up(self):
-        process = subprocess.Popen(self.subprocess_command, shell=True, stdout=subprocess.PIPE)
-        stdout = process.communicate()[0]
-
-        if not self.fp_status in stdout and not self.ros_loss_triggered:
-            self.connection_status.setText("Connection lost")
-            self.connection_status.setStyleSheet("color: red")
-            self.ros_loss_triggered = True
-        elif self.fp_status in stdout and self.ros_loss_triggered:
-            self.connection_status.setText("Connected")
-            self.connection_status.setStyleSheet("color: green")
-            self.ros_loss_triggered = False
+        # Set up core connection variables
+        self.core_clock_sub_name = rospy.get_param("/core_clock_sub","/mk3_core/clock")
+        self.core_clock_sub = rospy.Subscriber(self.core_clock_sub_name, Clock, self.core_clock_cb)
+        self.last_received_time = RosTime()
+
+        # Core timer
+        core_clock_pub_freq = rospy.get_param("~publish_frequency", 10)
+
+        #Set the frequency of the timer to be twice as fast as the clock frequency
+        freq = float(core_clock_pub_freq) * 2
+        self.core_clock_timer_dur = rospy.Duration(1.0 / freq)
+        self.core_clock_timer = rospy.Timer(self.core_clock_timer_dur, self.core_clock_timer_cb, oneshot = True)
+        self.poor_connection_threshold = rospy.get_param("/core_clock_poor_connection", 3)
+        self.lost_connection_threshold = rospy.get_param("/core_clock_lost_connection", 5)
+        self.first_value = False
+        self.get_first_msg = False
+        self.latency = 0
+        self.read_time = RosTime()
+        self.no_time_change = 0
+     
+        self.low_battery_popup.connect(self.bat_message_popup)
+
+
+    def core_clock_timer_cb(self, event):
+        self.core_clock_timer = rospy.Timer(self.core_clock_timer_dur, self.core_clock_timer_cb, oneshot = True)
+
+        if self.first_value:
```

Comment:
Could the self.first_value and self.get_first_msg flags be simplified out of this code?

What I'm thinking is that you could initialize self.last_received_time and self.read_time to None, and at the beginning of this callback you could do something like:

```
if read_time is None:
return (or set lost and return), we are still waiting for the first timer callback
elif last_received_time is None # read_time is not None since it's elif
last_received_time = last_read_time # works the same as the line below
return 
```

This should eliminate all the flags, unless I'm missing something. It should also allow to flatten the structure here quite a bit

Reply:
Yes I think you are correct. It works when i simulate the clock on my pc and run rr_viz. Will test it tomorrow when the robot is back together before pushing it here

Reply:
@msadowski I tried to implement it this way but it wasn't working so I think just to time constraints it's worth leaving it as it and make it a to do task later

Reply:
sure, no worries!